sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -618,6 +619,48 @@ object NullPropagation extends Rule[LogicalPlan] {
 }
 
 /**
+ * Propagate foldable expressions:
+ * Replace attributes with aliases of the original foldable expressions if possible.
+ * Other optimizations will take advantage of the propagated foldable expressions.
+ *
+ * {{{
+ *   SELECT 1.0 x, 'abc' y, Now() z ORDER BY x, y, 3
+ *   ==>  SELECT 1.0 x, 'abc' y, Now() z ORDER BY 1.0, 'abc', Now()
+ * }}}
+ */
+object FoldablePropagation extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = {
+    val foldableExprSet = ExpressionSet(plan.flatMap {
+      case Project(projectList, _) => projectList.collect {
+        case a: Alias if a.resolved && a.child.foldable => a
+      }
+      case _ => Nil
+    })
+
+    if (foldableExprSet.isEmpty) {
+      plan
+    } else {
+      val foldableMap = AttributeMap(foldableExprSet.toSeq.map(_.asInstanceOf[Alias])
+        .map(a => (a.toAttribute, a.child)))
```

Comment:
Okay, maybe there exists some problem if the logic is not clear to you. Let's consider the following example.
```
scala> sql("select 1, a newAlias from (select 1 a)").show
+---+--------+
|  1|newAlias|
+---+--------+
|  1|       1|
+---+--------+
```
The collected alias map is {'a':Literal(1)}.
And, we will replace `a newAliases` by creating new `Alias(Literal(1), "newAliases")`.
Does this make sense to you?